### PR TITLE
Add ATen pdist CPU kernel

### PR DESCRIPTION
**Pull request by erikbrinkman**
Also add single grad whitelist to the jit test

Test Plan:
```
python -m unittest test_torch.TestTorch.test_pdist_{empty,special,scipy}
test_nn.TestNN.test_pdist test_jit.TestJitGenerated.test_nn_pdist
```

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -2299,6 +2299,13 @@ def pairwise_distance(x1, x2, p=2, eps=1e-6, keepdim=False):
     return torch.pairwise_distance(x1, x2, p, eps, keepdim)
 
 
+def pdist(x, p=2):
+    r"""
+    See :func:`torch.pdist` for details
+    """
+    return torch.pdist(x, p)
```
Comment (ezyang) [inline torch/nn/functional.py]: If it's a direct pass through, you should just use `_add_docstr` to attach a docstring to the torch function you're binding. See https://github.com/pytorch/pytorch/blob/88025fe7776fbcfbbb87a0df1f7249348100020b/torch/nn/functional.py#L62 for an example.

File: test/test_jit.py [hunk 2]
```diff
@@ -6443,6 +6443,8 @@ def clone_inputs(requires_grad):
     self.assertEqual(grads, grads_test)
 
     # test the grad grad case
+    if self._testMethodName in nn_functional_single_grad:
+        return
```
Comment (ezyang) [inline test/test_jit.py]: Flagging @zdevito @jamesr66a. I find it hard to believe that this is the only function that only has single backwards (and not double backwards) implemented, that JIT script is exercising.

**Comment by ezyang** [comment]
How does the speed compare to scipy? ;)

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  static void run_parallel(Tensor& result, const Tensor& self, const double p) {
+    auto res_ = result.data<scalar_t>();
+    auto self_ = self.data<scalar_t>();
+    int64_t n = self.size(0);
+    int64_t m = self.size(1);
+    int64_t ns = self.stride(0);
+    int64_t ms = self.stride(1);
+
+    int64_t combs = n * (n - 1) / 2;
+    parallel_for(0, combs, 1, [=](int64_t k, int64_t end) {
```
Comment (ezyang) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: CC @cpuhrsch, you've got a user!

File: torch/nn/functional.py (hunk 1 above)
Comment (erikbrinkman) [reply torch/nn/functional.py]: torch pdist is a native function, i.e. defined in `aten/src/ATen/native/native_functions.yaml`, so it also doesn't have documentation. I'm happy to both add the pass through and documentation, but it's not clear what to do with native functions.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+      float n2 = n - .5;
+      // The -1 accounts for floating point truncation issues
+      int64_t i = (int64_t) ((n2 - std::sqrt(n2 * n2 - 2 * k - 1)));
```
Comment (ezyang) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: Nit: prefer `static_cast<int64_t>(...)`

File: torch/nn/functional.py (hunk 1 above)
Comment (ezyang) [reply torch/nn/functional.py]: I don't see documentation anywhere in the patch?

(Also, what's the plan for `squareform`?)

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: Also, bluuuugh, do we really have to do a `sqrt` to do index computation...

File: torch/nn/functional.py (hunk 1 above)
Comment (ezyang) [reply torch/nn/functional.py]: For native functions, the documentation is attached "post-facto" with `_add_docstr`. Check `torch/_tensor_docs.py`

**Comment by ezyang** [comment]
This looks very nice, and nicely parallelized. I'd say the main blocker is docs.

**Comment by erikbrinkman** [comment]
For timing:

Note that for an n x m input, space complexity is n^2, and time complexity is m n^2.

shape | scipy | torch
------|--------|-----
10 x 10 | 11.8 µs ± 280 ns | 6.65 µs ± 23.2 ns
10 x 1000 | 41.3 µs ± 900 ns | 9.41 µs ± 25.3 ns
1000 x 10 | 3.25 ms ± 94.8 µs | 311 µs ± 59.6 µs
1000 x 1000 | 241 ms ± 6.1 ms | 35.7 ms ± 88.7 µs
10000 x 1000 | 36.1 s ± 631 ms | 3.63 s ± 7.87 ms

File: torch/nn/functional.py (hunk 1 above)
Comment (erikbrinkman) [reply torch/nn/functional.py]: `squareform`?

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 4 above)
Comment (erikbrinkman) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: The only other solutions I can think of are to ahead of time:
```
vector<int64_t> lookup();
lookup.reserve(combs);
for (int64_t i = 0; i != n - 1; ++i) {
  for (int64_t j = i + 1; j != n; ++j) {
    lookup.push(i);
  }
}
```

and then later `int64_t i = lookup[k]`.

alternatively, we can ignore the startup time and memory for an O(n) computation:
```
int64_t kc = k;
int64_t i = 0;
for (int64_t kc = k, step = n - 1; kc >= step; i += 1, kc -= step, step -= 1) {}
```

This might be faster than the sqrt, but it's unclear. If you have a preference I can change it, but I doubt this is ever a bottleneck.

**Comment by cpuhrsch** [comment]
@erikbrinkman - Does you code benefit from using AVX2 or AVX? When you add a file to native/kernel it'll recompile and dispatch. But I don't see you make explicit use of those instructions. Did you time this to see whether it helps? (You can use the environment variables defined in ATen/native/DispatchStub.cpp. Sometimes, it's worse using those extended instruction sets. Also see "[Note SSE-AVX transitions]" if you see a significant slowdown (it might help explain it).

File: torch/nn/functional.py [hunk 5]
```diff
@@ -2299,6 +2299,24 @@ def pairwise_distance(x1, x2, p=2, eps=1e-6, keepdim=False):
     return torch.pairwise_distance(x1, x2, p, eps, keepdim)
 
 
+pdist = _add_docstr(torch.pdist, r"""
+pdist(input, p=2) -> Tensor
+
+Computes the p-norm distance between every pair of row vectors in the input.
```
Comment (gchanan) [inline torch/nn/functional.py]: Since this differs in the parameters from scipy (doesn't have a 'metric'), this should say something like "This is equivalent to scipy.spatial.distance.pdist with metric='minkowski'".

File: test/test_nn.py [hunk 6]
```diff
@@ -4821,6 +4821,20 @@ def test_pairwise_distance(self):
...
+            if trans:
+                inp = inp.transpose(0, 1)
+            for p in [0, 1, 2, 0.5, 1.5, 2.5, float('inf')]:
+                self.assertTrue(gradcheck(lambda x: F.pdist(x, p), (inp,)))
+
+    def test_pdist_zeros(self):
+        for trans in [False, True]:
+            inp = torch.randn(1, 3, requires_grad=True).repeat([2, 1])
+            for p in [0, 1, 2, 0.5, 1.5, 2.5, float('inf')]:
+                self.assertTrue(gradcheck(lambda x: F.pdist(x, p), (inp,)))
```
Comment (gchanan) [inline test/test_nn.py]: nit: In situations like this I find it helpful to throw in an expected failure case when gradgradcheck is called.  This means if someone fixes double backwards they have to remember to test it.

File: test/test_torch.py [hunk 7]
```diff
@@ -1043,6 +1043,56 @@ def test_pairwise_distance_empty(self):
...
+                self.assertEqual(expected, actual)
+
+    @skipIfRocm
+    @unittest.skipIf(not TEST_SCIPY, "Scipy not found")
+    def test_pdist_scipy(self):
+        from scipy.spatial.distance import pdist
+        devices = ['cpu']
+        for device in devices:
+            for shape in [(4, 5), (3, 2), (2, 1)]:
+                for p in [1, 2, 3, 1.5, 2.5]:
```
Comment (gchanan) [inline test/test_torch.py]: these don't seem to match up when p is inf -- is that a bug in scipy?

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+  static scalar_t pdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += std::pow(std::abs(*a - *b), p);
+    }
+    return std::pow(result, 1.0 / p);
+  }
+
+  static scalar_t idist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: this gives the same answer for inf and -inf -- is that right?  It's a little unclear to me because the scipy docs say p >= 1, but it accepts other values...

File: torch/nn/functional.py (hunk 5 above)
Comment (erikbrinkman) [reply torch/nn/functional.py]: I considered this, but I wondered how much of a reference you wanted to have to scipy in the docs. It might become invalid as scipy changes, and as you point out below, it doesn't always map to the "right" value. I'm going to assume that you still want me to go ahead with this change unless noted otherwise.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 8 above)
Comment (erikbrinkman) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: Yeah. I need to add a check that p must be in [0, inf]

File: test/test_torch.py (hunk 7 above)
Comment (erikbrinkman) [reply test/test_torch.py]: The doc for pdist just says >= 1 and passes a float. The backend `minkowski` function specifically says an integer, but clearly works in (0, inf). For 0 it returns inf, for inf it returns 1.

File: torch/nn/functional.py (hunk 1 above)
Comment (erikbrinkman) [reply torch/nn/functional.py]: Do you mean adding a `squareform` function? It wouldn't be bad, but it probably makes more sense to add something equivalent to numpy's {triu,tril}_indices. They'll allow easy conversion and are a tad more general.

File: torch/nn/functional.py (hunk 5 above)
Comment (gchanan) [reply torch/nn/functional.py]: Good question!  I think in this case it is justified because the function has the exact same name and we are talking about parameter names, which would have to change in a backwards-incompatible way for our docs to be invalid.

File: test/test_torch.py (hunk 7 above)
Comment (gchanan) [reply test/test_torch.py]: That's pretty strange on their side.  So, what do you think the correct thing to do here is?  Add your own check for 0 and inf that doesn't go through scipy (assuming that your implementation gives different results)?

Comment (erikbrinkman) [reply test/test_torch.py]: Yeah, that's why there's `test_pdist_special` which tests the "special" values of 0 and infinity.

Comment (gchanan) [reply test/test_torch.py]: Ah, okay I didn't know what "special" meant.  Maybe a comment (or moving it after the normal tests) would help clarify.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
+#include "ATen/native/cpu/ReduceOpsKernel.h"
```
Comment (SsnL) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: If you are not using vectorized CPU instructions, please do not write your kernel in `native/cpu`.

File: torch/nn/functional.py [hunk 10]
```diff
@@ -2299,6 +2299,30 @@ def pairwise_distance(x1, x2, p=2, eps=1e-6, keepdim=False):
...
+This is identical to the upper triangular portion, excluding the diagonal, of
+`torch.norm(input[:, None] - input, dim=2, p=p)`. This function will be faster
+if the rows are contiguous.
+
+If input has shape :math:`N \times M` then the output will have shape
+:math:`\frac{1}{2} N (N - 1)`.
```
Comment (SsnL) [inline torch/nn/functional.py]: You are missing a \times M here.

Comment (erikbrinkman) [reply torch/nn/functional.py]: Where?

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 9 above)
Comment (gchanan) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: We should probably change this globally, maybe moving native/cpu to native/cpu/intrisic_ops or something.  It's confusing that CPU-only and code that works across CPU/CUDA unchanged both go in native.

Comment (gchanan) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: @cpuhrsch?

File: torch/nn/functional.py (hunk 10 above)
Comment (SsnL) [reply torch/nn/functional.py]: Oh my bad. I thought it returns a 2d tensor. Sorry about it!

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 9 above)
Comment (SsnL) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: @erikbrinkman Yes, please move it into native.

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: @erikbrinkman - let's sync up on this offline. We had talked about whether you want to use vectorized instructions or not. Aside from this, I'll update the README within this folder and we'll probably want to move things or rename them just as greg suggested.

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: We just talked about this in person and the conclusion is that it's worth benchmarking whether or not some level of auto-vectorization is beneficial here. If it's not, this can be moved back into native.

File: torch/nn/functional.py (hunk 10 above)
Comment (gchanan) [reply torch/nn/functional.py]: Might make it clearer if you print it as a tuple (e.g. (n/(n-1),)

Comment (erikbrinkman) [reply torch/nn/functional.py]: I was generally confused when reading documentation. Is math or cs notation preferred? I.e. should I replace all shapes with tuples e.g. `(n * (n - 1) // 2,)`?

File: test/test_torch.py [hunk 11]
```diff
@@ -1043,6 +1043,45 @@ def test_pairwise_distance_empty(self):
             self.assertEqual(torch.zeros(0, device=device), torch.pairwise_distance(x, y))
             self.assertEqual(torch.zeros((0, 1), device=device), torch.pairwise_distance(x, y, keepdim=True))
 
+    @skipIfRocm
```
Comment (ezyang) [inline test/test_torch.py]: Huh, does this really fail on ROCM? That's really strange!

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: It's fine.

**Review by ezyang — APPROVED** [review]
But don't land until tests pass...
